src/phongo_bson_encode.c

```diff
@@ -128,20 +177,15 @@ static void php_phongo_bson_append_object(bson_t* bson, php_phongo_field_path* f
 				return;
 			}
 
-			if (Z_TYPE(obj_data) != IS_ARRAY && !(Z_TYPE(obj_data) == IS_OBJECT && instanceof_function(Z_OBJCE(obj_data), zend_standard_class_def))) {
-				phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE,
-									   "Expected %s::%s() to return an array or stdClass, %s given",
-									   ZSTR_VAL(Z_OBJCE_P(object)->name),
-									   BSON_SERIALIZE_FUNC_NAME,
-									   PHONGO_ZVAL_CLASS_OR_TYPE_NAME(obj_data));
+			if (!phongo_check_bson_serialize_return_type(&obj_data, Z_OBJCE_P(object))) {
+				// Exception already thrown
```

Comment:
@alcaeus: I see this was already merged, but note that `phongo_check_bson_serialize_return_type()` can technically return `false` without throwing an exception if the object is neither Serializable nor Persistable. Perhaps impossible in practice, but I wanted to point it out since this seems like undefined behavior.

Reply:
Good point - taken care of in https://github.com/mongodb/mongo-php-driver/pull/1453